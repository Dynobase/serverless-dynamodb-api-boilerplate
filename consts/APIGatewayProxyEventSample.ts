import { APIGatewayProxyEvent } from "aws-lambda";

export const APIGatewayProxyEventSample: APIGatewayProxyEvent = {
  body: ``,
  headers: {},
  httpMethod: "POST",
  isBase64Encoded: false,
  path: "/",
  queryStringParameters: {},
  pathParameters: {},
  stageVariables: {},
  multiValueHeaders: {},
  requestContext: null,
  resource: "/",
  multiValueQueryStringParameters: {}
};
