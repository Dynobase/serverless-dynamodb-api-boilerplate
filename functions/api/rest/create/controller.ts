import { APIGatewayProxyEvent } from "aws-lambda";
import { DynamoDB } from "aws-sdk";
import { uuid } from "uuidv4";
import { ThingModel } from "../../../../models/Thing";
import { Thing } from "./request_schema";

// Separated business logic with injectable documentClient for testing purposes
export const createController = async (
  event: APIGatewayProxyEvent,
  documentClient: DynamoDB.DocumentClient
) => {
  // Parse string to verified request schema model
  const input: Thing = JSON.parse(event.body);
  const item = {
    id: uuid(),
    status: 'OK',
    date_added: (+new Date()).toString(),
    name: input.name,
  };

  const params = ThingModel.put(item);

  console.log(params);

  const result = await documentClient.put(params).promise();

  return result.Attributes;
};
