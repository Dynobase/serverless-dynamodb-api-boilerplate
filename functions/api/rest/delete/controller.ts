import { APIGatewayProxyEvent } from "aws-lambda";
import { DynamoDB } from "aws-sdk";
import { ThingModel } from "../../../../models/Thing";

// Separated business logic with injectable documentClient for testing purposes
export const deleteController = async (
  event: APIGatewayProxyEvent,
  documentClient: DynamoDB.DocumentClient
) => {
  let item = {
    id: event.queryStringParameters.id,
    status: event.queryStringParameters.status,
    date_added: event.queryStringParameters.date_added,
  }
  
  const params = ThingModel.delete(item)
  const response = await documentClient.delete(params).promise()
  
  return ThingModel.parse(response);
};
