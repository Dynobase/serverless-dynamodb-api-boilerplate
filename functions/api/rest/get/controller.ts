import { APIGatewayProxyEvent } from "aws-lambda";
import { DynamoDB } from "aws-sdk";
import { ThingModel } from "../../../../models/Thing";

// Separated business logic with injectable documentClient for testing purposes
export const getController = async (
  event: APIGatewayProxyEvent,
  documentClient: DynamoDB.DocumentClient
) => {
  let item = {
    id: event.queryStringParameters.id,
    status: event.queryStringParameters.status,
    date_added: event.queryStringParameters.date_added,
  }
  
  const params = ThingModel.get(item)
  const response = await documentClient.get(params).promise()
  
  return ThingModel.parse(response);
};
