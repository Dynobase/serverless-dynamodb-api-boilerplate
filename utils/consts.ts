export const getEndpoint = () => {
  if (process.env.AWS_EXECUTION_ENV) {
    return undefined;
  } else if (process.env.DYNAMODB_ENDPOINT) {
    return `http://${process.env.DYNAMODB_ENDPOINT}`;
  } else {
    return "http://localhost:8000";
  }
};

export const getRegion = () => {
  if (process.env.AWS_EXECUTION_ENV) {
    return undefined;
  } else if (process.env.AWS_REGION) {
    return process.env.AWS_REGION;
  } else if (process.env.AWS_DEFAULT_REGION) {
    return process.env.AWS_DEFAULT_REGION;
  } else {
    return "us-east-1";
  }
};

export default () => process.env.DATA_STORE_ARN.split("/").slice(-1)[0];
