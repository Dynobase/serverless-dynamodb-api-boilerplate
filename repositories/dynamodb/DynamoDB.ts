import * as AWSXRay from "aws-xray-sdk";
import { DynamoDB } from "aws-sdk";
import { getEndpoint, getRegion } from "../../utils/consts";

export const documentClient = new DynamoDB.DocumentClient({
  service: new DynamoDB({
    endpoint: getEndpoint(),
    region: getRegion()
  })
});

// Capture X-Ray traces only on AWS Lambda
if (process.env.AWS_EXECUTION_ENV) {
  AWSXRay.captureAWSClient((documentClient as any).service);
}
