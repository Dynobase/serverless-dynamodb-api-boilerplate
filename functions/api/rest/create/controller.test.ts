import { createController } from "./controller";
import { documentClient } from "../../../../services/DynamoDB";
import { APIGatewayProxyEventSample } from "../../../../consts/APIGatewayProxyEventSample";

describe("REST/Create", () => {
  test("returns created Thing", async done => {
    const response = await createController(
      {
        ...APIGatewayProxyEventSample,
        body: `{"name":"Testing"}`
      },
      documentClient // can be mocked to disable DB access
    );

    expect(response).toMatchSnapshot();
    done();
  });
});
