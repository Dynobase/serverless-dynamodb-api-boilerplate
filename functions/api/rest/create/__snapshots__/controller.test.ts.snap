// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REST/Create returns created Thing 1`] = `
Object {
  "date_added": "1584184460879",
  "id": "f328ee8b-7e8a-4e8e-9525-fb1e42a11491",
  "name": "Testing",
  "status": "OK",
}
`;
